otter/controller.py

```diff
@@ -348,3 +371,88 @@ def calculate_delta(log, state, config, policy):
+    """
+    try:
+        server_info = yield retry_effect(get_server_details(server_id),
+                                         retry_times(3),
+                                         exponential_backoff_interval(2))
+    except NoSuchServerError:
+        raise ServerNotFoundError(group.tenant_id, group.uuid, server_id)
+
+    group_id = group_id_from_metadata(
+        get_in(('server', 'metadata'), server_info, {}))
+    if group_id != group.uuid:
+        raise ServerNotFoundError(group.tenant_id, group.uuid, server_id)
+
+
+@do
+def _can_maybe_scale_down(group, server_id):
+    """
+    Given a group and a server ID, determines if the group can be scaled down.
+    If not, it raises a :class:`CannotDeleteServerBelowMinError`.
+    """
+    manifest = yield Effect(GetScalingGroupInfo(
+        tenant_id=group.tenant_id, group_id=group.uuid))
+    min_entities = manifest['groupConfiguration']['minEntities']
+    state = manifest['state']
+
+    if state.desired == min_entities:
+        raise CannotDeleteServerBelowMinError(
+            group.tenant_id, group.uuid, server_id, min_entities)
+
+
+@do
+def convergence_remove_server_from_group(
+        group, state, server_id, replace, purge):
+    """
+    Remove a specific server from the group, optionally decrementing the
+    desired capacity.
+
+    The server may just be scheduled for deletion, or it may be evicted from
+    the group by removing otter-specific metdata from the server.
+
+    :param group: The scaling group to remove a server from.
+    :type group: :class:`~otter.models.interface.IScalingGroup`
+    :param bytes server_id: The id of the server to be removed.
+    :param bool replace: Should the server be replaced?
+    :param bool purge: Should the server be deleted from Nova?
+
+    :return: The updated state.
+    :rtype: deferred :class:`~otter.models.interface.GroupState`
+
+    :raise: :class:`CannotDeleteServerBelowMinError` if the server cannot
+        be deleted without replacement, and :class:`ServerNotFoundError` if
+        there is no such server to be deleted.
+    """
+    effects = [_is_server_in_group(group, server_id)]
+    if not replace:
+        effects.append(_can_maybe_scale_down(group, server_id))
+
+    # the (possibly) two checks can happen in parallel, but we want
+    # ServerNotFoundError to take precedence over
+    # CannotDeleteServerBelowMinError
+    both_checks = yield parallel_all_errors(effects)
+    for is_error, result in both_checks:
+        if is_error:
+            reraise(*result)
+
+    # Remove the server
+    if purge:
+        eff = set_nova_metadata_item(server_id, *DRAINING_METADATA)
+    else:
+        eff = Effect(
+            EvictServerFromScalingGroup(scaling_group=group,
+                                        server_id=server_id))
+    yield retry_effect(eff, retry_times(3), exponential_backoff_interval(2))
+
+    if not replace:
+        state.desired -= 1
```

Comment:
I wouldn't mention this in "normal" code, but since this is an Effect-returning function: this is a side-effect that's not encapsulated in an Effect. We have `ModifyGroupState` for changing groups, which might be suitable here, but given that you're returning `state` I'm not sure what your plans are (sorry I haven't read your other in-progress branch yet). You could also return a *new* state object -- you can just return `assoc_obj(state, desired=state.desired - 1)`.